core/wasmWasi/src/files/FileSystemWasm.kt

```diff
@@ -5,36 +5,539 @@
+        preopens.map { it.path }
+    }
+
+    internal val preopens: List<PreOpen> by lazy {
+        // 0 - stdin, 1 - stdout, 2 - stderr, 3 - if the first preopened directory, if any
+        val firstPreopenFd = 3
+        val rootPaths = mutableListOf<PreOpen>()
+
+        for (fd in firstPreopenFd..<Int.MAX_VALUE) {
+            if (!loadPreopenInfo(fd, rootPaths)) {
+                break
+            }
+        }
+
+        rootPaths
+    }
+
+    internal fun getRootOrNull(path: Path): PreOpen? {
+        if (!path.isAbsolute) {
+            return preopens.firstOrNull()
+        }
+
+        for (root in preopens) {
+            var p: Path? = path
+            while (p != null) {
+                if (root.path == p) return root
+                p = p.parent
+            }
+        }
+        return null
+    }
+
+    internal fun getRoot(path: Path): PreOpen {
+        return getRootOrNull(path) ?: throw IOException("Path does not belong to any preopened directory: $path")
+    }
+
+    @OptIn(UnsafeWasmMemoryApi::class)
+    private fun loadPreopenInfo(fd: Int, outputList: MutableList<PreOpen>): Boolean {
+        return withScopedMemoryAllocator { allocator ->
+            val prestat = Prestat(allocator)
+            val res = Errno(fd_prestat_get(fd, prestat.address))
+
+            if (res == Errno.badf) {
+                return@withScopedMemoryAllocator false
+            }
+
+            if (res != Errno.success) {
+                throw IOException("Unable to process fd=$fd as preopen: ${res.description}")
+            }
+
+            check(prestat.type == PrestatType.dir) { "Unsupported prestat type" }
+
+            val nameLength = prestat.nameLength
+            val pathBuffer = allocator.allocate(nameLength)
+            val dirnameRes = Errno(fd_prestat_dir_name(fd, pathBuffer.address.toInt(), nameLength))
+            if (dirnameRes != Errno.success) {
+                throw IOException("Unable to get preopen dir name for fd=$fd: ${dirnameRes.description}")
+            }
+            outputList.add(PreOpen(Path(pathBuffer.loadBytes(nameLength).decodeToString()), fd))
+
+            true
+        }
+    }
+}
+
+private const val TEMP_CIOVEC_BUFFER_LEN = 8192
+
+private class FileSink(private val fd: Fd) : RawSink {
+    private var closed: Boolean = false
+
+    @OptIn(UnsafeWasmMemoryApi::class)
+    override fun write(source: Buffer, byteCount: Long) {
+        check(byteCount >= 0)
+        if (byteCount == 0L) return
+        withScopedMemoryAllocator { allocator ->
+            val buffer = allocator.allocate(TEMP_CIOVEC_BUFFER_LEN)
+            var remaining = byteCount
+
+            val ciovec = Ciovec(allocator).also {
+                it.buffer = buffer
+            }
+
+            val resultPtr = allocator.allocate(4)
+
+            while (remaining > 0) {
+                val toRead = minOf(remaining, TEMP_CIOVEC_BUFFER_LEN).toInt()
+                source.readToLinearMemory(buffer, toRead)
```

Comment:
It is not immediately clear which buffer was meant. Perhaps we should call it something else?